### PR TITLE
Add support for arbitrary number of search providers.

### PR DESCRIPTION
**Pull request by jahands**
This change is backwards compatible with the old config.
This way you can add things like youtube/wikipedia/github/google/bing/yahoo/anything/etc to the providers.

### CODE REVIEW

**Comment by jahands** [comment]
New config looks like this:

```
{
    "codebing.searchproviders": {
        "defaultSearchProviderID": "g",
        "b": "https://www.bing.com/search?q={query}",
        "g": "https://www.google.com/search?q={query}"
    }
}
```

Use like this:
- `ctrl+alt+f`
- `g This is a google search!`
- `b This is a bing search!`
- `This is a search using the default provider!`

File: package.json [hunk 1]
```diff
@@ -30,10 +30,27 @@
...
       "type": "object",
       "properties": {
-        "codebing.searchprovider": {
-          "type": "string",
-          "default": "https://www.bing.com/search?q={query}",
-          "description": "A string to an online search engine. {query} is replaced with the user search."
+        "codebing.searchproviders": {
+          "title": "Search providers",
+          "type": "object",
+          "description": "A string to an online search engine. {query} is replaced with the user search. First word is used as selecter (b or g by default).",
```
Comment (SrTobi) [inline package.json]: Should be something like: `A dictionary with search engines, mapping a key to a search url. In the search url {query} is replaced with the user search. Use the key as first word in the search input to specify the search engine.`

File: package.json [hunk 2]
```diff
@@ -30,10 +30,27 @@
...
+          "default": {
+            "defaultSearchProviderID": "b",
```
Comment (SrTobi) [inline package.json]: I think this should be a separate option.

File: src/extension.ts [hunk 3]
```diff
@@ -41,16 +41,41 @@ export function activate(context: vscode.ExtensionContext) {
...
 
-function searchfor(query:string) {
+function searchfor(query: string) {
 	// Build the query
 	let config = vscode.workspace.getConfiguration("codebing");
-	let searchprovider = config.get("searchprovider") as string;
-	let q = query.replace(/[\r\n]/g, "");
-	let url = searchprovider.replace("{query}", q);
+	let searchProviders = config.get("searchproviders") as {};
+	let defaultOnly = config.get("defaultprovideronly") as string;
```
Comment (SrTobi) [inline src/extension.ts]: should be `as boolean`... or i think it is done like `config.get<boolean>(...)`. See [here](https://code.visualstudio.com/docs/extensionAPI/vscode-api#WorkspaceConfiguration). I saw that somewhere else as well.

File: src/extension.ts [hunk 4]
```diff
@@ -41,16 +41,41 @@ export function activate(context: vscode.ExtensionContext) {
...
 }
 
-function searchfor(query:string) {
+function searchfor(query: string) {
 	// Build the query
 	let config = vscode.workspace.getConfiguration("codebing");
-	let searchprovider = config.get("searchprovider") as string;
-	let q = query.replace(/[\r\n]/g, "");
-	let url = searchprovider.replace("{query}", q);
+	let searchProviders = config.get("searchproviders") as {};
```
Comment (SrTobi) [inline src/extension.ts]: you can use `as { [id: string]: string; };`

File: src/extension.ts [hunk 5]
```diff
@@ -41,16 +41,41 @@ export function activate(context: vscode.ExtensionContext) {
...
+	if (oldSearchProvider != null) {
+		searchProviders["defaultSearchProviderID"] = "oldSearchProvider";
+		searchProviders["oldSearchProvider"] = oldSearchProvider;
+	}
 	
-	open(url);
+	// Get the first word to use as the search provider ID
+	let searchProviderID = query.split(' ', 1)[0];
+	// Retrieve the search provider based on the ID
+	let searchProvider = searchProviders[searchProviderID]
```
Comment (SrTobi) [inline src/extension.ts]: ;

File: src/extension.ts [hunk 6]
```diff
@@ -41,16 +41,41 @@ export function activate(context: vscode.ExtensionContext) {
...
+	let q = "" // Query variable
```
Comment (SrTobi) [inline src/extension.ts]: ;

**Comment by SrTobi** [comment]
Very good idea. There are also some minor errors like missing `;`.

**Comment by jahands** [comment]
Thanks for the info! Refactored to use a cleaner config.

**Pull request merged**